### PR TITLE
Register dynamic processor setting.

### LINKED ISSUE #3443: Dynamic Dropdown Options in Processor Settings
Add the ability to use a processor setting that supports a dynamic set of options, ie newsletter lists.

- [ ]  Options should be loaded by a callback function.
- - [ ] ? Options may be loaded by a provided REST API endpoint.
- [ ]  Fetched options should be shared between multiple processors.
- [ ]  Cache options to limit redundant network requests.
- [ ]  Bust cache for fresh data.
- [ ]  Include a loading animation when fetching options.

### PR DESCRIPTION
**Pull request by kjohnson**
Closes #3443.

Registers a new "dynamic" processor setting type, which uses a user-defined callback to load options.

- Adds a new field type `dynamic`.
- Accepts a new config item `callback`.
- Passes a Promise to resolve fetched select options.
- Supports an array of options as a default state without an initial network request.

Example Use: https://gist.github.com/kjohnson/0a5e556e8b279537f17227eef42f3a3a

### CODE REVIEW

**Comment by Shelob9** [comment]
@kjohnson This is submitted against the wrong branch. I editted it here, but beacuse you forked off of master, we have extra stuff which may or may not merge fine, we'll see once the #3442  and #3441 PRs are merged. In the future, please follow the workflow in the contibutor's guide: https://github.com/CalderaWP/Caldera-Forms/blob/master/contributing/pull-request-workflow.md

File: processors/classes/ui.php [hunk 1]
```diff
@@ -347,6 +350,21 @@ public static function input( $type, $args, $id, $classes, $required, $has_desc
 					'{{' . esc_attr( $id ) . '}}'
 				);
 				break;
+			case 'dynamic':
+				$field = sprintf('<select id="{{_name}}[%s]" class="%s dynamic" name="{{_name}}[%s]" data-value="{{%s}}" data-callback="%s"></select>',
```
Comment (Shelob9) [inline processors/classes/ui.php]: This addition makes the inline docs for this method inaccurate. Please update them as well.

File: assets/js/edit.js [hunk 2]
```diff
@@ -2740,4 +2740,51 @@ function cf_revisions_ui() {
...
 }
 
+/**
+ * Processor Dynamic Options
+ * @TODO Move this wherever it needs to go.
+ */
+
+window.cfProcessorDynamicOptionsCache = {};
+
+function cfProcessorDynamicOptionsLoad(id) {
```
Comment (Shelob9) [inline assets/js/edit.js]: This function is missing inline documentation and tests.

File: assets/js/edit.js [hunk 3]
```diff
@@ -2740,4 +2740,51 @@ function cf_revisions_ui() {
...
+    if(window.cfProcessorDynamicOptionsCache[callback]) {
+        console.log('Update option from cache');
+        cfProcessorDynamicOptionsUpdate(select, window.cfProcessorDynamicOptionsCache[callback]);
+    } else {
+        console.log('Fetching options');
+        cfProcessorDynamicOptionsFetch(id);
+    }
+}
+
+function cfProcessorDynamicOptionsRefresh(id) {
```
Comment (Shelob9) [inline assets/js/edit.js]: This function is missing inline documentation and tests.

File: assets/js/edit.js [hunk 4]
```diff
@@ -2740,4 +2740,51 @@ function cf_revisions_ui() {
...
+    const select = document.getElementById(id);
+    select.options.length = 0;
+    cfProcessorDynamicOptionsFetch(id)
+}
+
+function cfProcessorDynamicOptionsFetch(id) {
```
Comment (Shelob9) [inline assets/js/edit.js]: This function is missing inline documentation and tests.

File: assets/js/edit.js [hunk 5]
```diff
@@ -2740,4 +2740,51 @@ function cf_revisions_ui() {
...
+    select.classList.add('dynamic-fetching');
+    const callback = select.getAttribute('data-callback');
+    const promise = new Promise(window[callback]);
+    promise.then(function(options) {
+        window.cfProcessorDynamicOptionsCache[callback] = options;
+        cfProcessorDynamicOptionsUpdate(select, options);
+    });
+}
 
+function cfProcessorDynamicOptionsUpdate(select, options) {
```
Comment (Shelob9) [inline assets/js/edit.js]: This function is missing inline documentation and tests.

File: assets/js/edit.js [hunk 6]
```diff
@@ -2740,4 +2740,51 @@ function cf_revisions_ui() {
...
+    const value = select.getAttribute('data-value');
+    select.options.length = 0;
+    options.map(function(option) {
```
Comment (Shelob9) [inline assets/js/edit.js]: Please make sure to add tests to cover then `options` is an empty array and for when it is undefined.

File: assets/js/edit.js [hunk 7]
```diff
@@ -2740,4 +2740,51 @@ function cf_revisions_ui() {
...
+}
+
+function cfProcessorDynamicOptionsFetch(id) {
+    const select = document.getElementById(id);
+    select.classList.add('dynamic-fetching');
+    const callback = select.getAttribute('data-callback');
+    const promise = new Promise(window[callback]);
+    promise.then(function(options) {
+        window.cfProcessorDynamicOptionsCache[callback] = options;
+        cfProcessorDynamicOptionsUpdate(select, options);
```
Comment (Shelob9) [inline assets/js/edit.js]: Please validate if select is an element before using it.

**Comment by kjohnson** [comment]
@Shelob9 I still need to add the tests, but it looks like I can't run Jest tests on `assets/js/edit.js` due to the file itself not being in a testable state.

We should discuss where this JS should live so that I can add in the unit tests for the functions.

**Comment by Shelob9** [comment]
> @Shelob9 I still need to add the tests, but it looks like I can't run Jest tests on assets/js/edit.js due to the file itself not being in a testable state.

Not sure I agree with using unit tests for this. I think we're talking about something that is tocuhed/ viewed by the user, also it involves window/document objects. I think that cypress.io tests -- see #3441 3438

If you want to use a unit test, move the function to https://github.com/CalderaWP/Caldera-Forms/tree/master/clients/functions and require it in edit.js, but I don't see the need for that extra complexity.

**Comment by kjohnson** [comment]
@Shelob9 so when you noted that the functions weren't covered by tests, you meant that e2e tests meet that requirement?

**Comment by Shelob9** [comment]
> @Shelob9 so when you noted that the functions weren't covered by tests, you meant that e2e tests meet that requirement?

I meant there were not any tests. I should have added thaughts about what kind of tests. That's on me.
